Additional features

leiflm opened the pull request:
This series of commit adds many features one might want to get the most out of one's UML diagram.

leiflm committed:
Adds the following features:

* support for packages
* handling of instance objects
* handling of methods (FunctionBehavior)
* import of referenced Variables
* Constructor creation
* many more small things

leiflm committed:
use OrderedSet within `for`-iterators to avoid random reordering

leiflm committed:
Match UML's `ImportElement` notation with Python's import statement

Beware, that a certain import order might be necessary. That's why custom imports
are prepended.

leiflm committed:
generalize protected import-code section generation

leiflm committed:
do not generate parameter `self` for static methods

leiflm committed:
fixup of 8a557014e99ac2f61ca8b2a7ef459ef1415cebb0

The previous commit did not handle certain function decleration cases correctly.

leiflm committed:
adds support for UML <<use>> (Usage) relations.

Whenever a class uses another class, a matching `import` stmt is generated.
This is particularly usefule for UML function definitions, that invoke other
(static classes') functions.

leiflm committed:
Supercharge support for <<use>> by supporting any kind of Dependency.

This is an amendment to 518ed2aa0a0b0bd648b762a540e98909c51adb2f.

leiflm committed:
reduces blank lines during default value code generation

leiflm committed:
improve whitespacing

leiflm committed:
fixes element order for imports, etc.

OCL `asOrderedSet()` is only a typecast. Use `sortedBy` to actually sort elements (by name).

leiflm committed:
enables the user to effectively override Operation bodies.

Moves protected section to the begging of an Operation's body

leiflm committed:
only sort otherwise unsorted elements alphabetically (i.e. inheritance)

leiflm committed:
reworks UML default value implications. Refactor generation code

Before this change, if a variable'd default value was undefined,
the generate would generate code to instanciate a class instance.

The behavior interfered with code generation for method/function parameters.

leiflm committed:
adds handling of instances inside InstanceValues

before, it only derived the type to invoke a constructor.
The behavior did not work for enumerations.

leiflm committed:
fixes LiteralBoolean handling

leiflm committed:
use Python's `@staticmethod` decorator instead of `@classmethod` for static methods

According to Python's[0] documentation, they are more similar to Java/C++'s
static methods.

[0] https://docs.python.org/3.6/library/functions.html#staticmethod

leiflm committed:
add some type safety to functions too

Following mypy declaration style [0] to add some more type safety.

[0] http://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#functions

leiflm committed:
add function parameter type safety for PrimitiveTypes

Complex Types are a bit more hard to handle, since one needs to
iterate all parameters of all operations to generate proper include
statements. This is postponed until a refactoring of the import
statement code generation.

Also fixup of ab6a4718e920cd89bba2ff157949769b44a311d0

leiflm committed:
import ABC stuff to declare abstract structures properly

leiflm committed:
update README.md to document the added features.

leiflm committed:
remove unused MTL file

leiflm committed:
README.md: add comment about inheritance statement order too

aranega commented:
That is just awesome! You really went from the simple script it was to a really greeat one. 

Thanks a lot @leiflm!

leiflm commented:
my pleasure. I think I could/should have:

- Added more comments about specific aspects (like handling of `PrimitiveType`)
- Refactored the code (I learned how mtl works, during extension of your script)

Anyway. I hope others might benefit from this work too.

aranega commented:
I'm sure that other people will benefit from your work! Regarding further modifications, they are really welcome, but don't worry about it, your contribution is already so great. I quickly checked your code that sounds good to me. There is some parts where templat with guards could have been used in order to split a big template in many others (for example, the `header` template in order to separate static/abstract/normal case), but that's not so much an issue.

I'm also working on a pure python implementation of Ecore and UML (named PyEcore and PyUML2), this is possible that in the future I "convert" this Accelo script in pure Python + templates in order to provide a full command line generator. I will take inspiration from the modifications you made in order to transform in the same way the whole input UML file.
